src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,956 @@
+                            }
+
+                            if (!IsEmptyVersion(info.Version))
+                            {
+                                capturedVersions[info.PlatformName] = info.Version;
+                            }
+                        }
+                        else
+                        {
+                            capturedPlatforms.Add(info.PlatformName);
+
+                            if (capturedVersions.Any())
+                            {
+                                if (attribute.UnsupportedFirst != null && capturedVersions.TryGetValue(info.PlatformName, out var version) && attribute.UnsupportedFirst >= version)
+                                {
+                                    attribute.UnsupportedFirst = null;
+                                }
+
+                                if (attribute.Obsoleted != null && capturedVersions.TryGetValue(info.PlatformName, out version) && attribute.Obsoleted <= version)
+                                {
+                                    attribute.Obsoleted = null;
+                                }
+
+                                if (attribute.UnsupportedSecond != null && capturedVersions.TryGetValue(info.PlatformName, out version) && attribute.UnsupportedSecond <= version)
+                                {
+                                    attribute.UnsupportedSecond = null;
+                                }
+                            }
+
+                            if (attribute.SupportedFirst != null && attribute.SupportedFirst <= info.Version)
+                            {
+                                attribute.SupportedFirst = null;
+                            }
+
+                            if (attribute.SupportedSecond != null && attribute.SupportedSecond <= info.Version)
+                            {
+                                attribute.SupportedSecond = null;
+                            }
+                        }
+                    }
+                }
+
+                if (value.Parents.IsEmpty)
+                {
+                    foreach (var attribute in attributes)
+                    {
+                        // if any of the attributes is not suppressed
+                        if (attribute.Value.HasAttribute())
+                        {
+                            return false;
+                        }
+                    }
+                }
+                else
+                {
+                    // 'GlobalFlowStateAnalysisValueSet.Parents' represent || of values on different flow paths.
+                    // We are guarded only if values are guarded on *all flow paths**.
+                    foreach (var parent in value.Parents)
+                    {
+                        // NOTE: IsKnownValueGuarded mutates the input values, so we pass in cloned values
+                        // to ensure that evaluation of each part of || is independent of evaluation of other parts.
```

Comment:
Mutation seems problematic because the analysis runs concurrently. Also, all this cloning seems error prone. Have we considered an immutable approach?

Reply:
+1 I would also like PlatformAttributes data structure to be written as a readonly struct after 5.0 as part of generic code cleanup. Current mutable implementation makes it hard to maintain and code is hard to understand. @buyaa-n made sure the mutations happens on the code path where a single thread is running, but it is easy to see someone else violate that in future.

Reply:
Cool--after 5.0 sounds good